### ISSUE TITLE
NPM package with plain JS API

### ISSUE THREAD

**Issue by martinklepsch** [kkinnear/zprint#45]
For integration with JS-based editors it would be nice if there was an NPM package exposing a standard JS API (CommonJS?). I understand if this is out of scope for this project but wanted to leave a note here just in case :) Thank you! 

Edit: An NPM package could also expose a binary similar to `lumo` which could help with overall startup time.

Edit Jan 15th '18: Here's a [twitter conversation](https://twitter.com/shaunlebron/status/952859439354339330) that may provide additional context and (I think) stimulated recent activity around this issue.

**Comment by kkinnear** [comment]
I'd be delighted to be able to offer an NPM package, but unfortunately I don't have the first idea of how to do so.  If you want to point me to some how-to information that would let someone who knows little to nothing about NPM figure out how to build such a package, I'd certainly be willing to give it a look.  As to creating a standard JS API, I'm equally in the dark on that one too.  Again, if you want to propose something, I'm certainly open to suggestions.  

All that said, if you know how to do this, and you know what you think would be useful to yourself (presumably) as well as to other folks, I would be delighted to have you dive in an create one or more pull requests to make all of this happen.  Someone would have to release a new version whenever zprint changes (or at least every few changes).  If it wasn't too onerous, I would be willing to do that going forward, but I doubt I'm going to be the one to figure out how to do this from scratch.

**Comment by roman01la** [comment]
Hey. What's the status on this issue? Is any help needed? Or did you decide to use something else? @martinklepsch

**Comment by kkinnear** [comment]
I'd love your help!  I don't even know how to make an NPM package, let alone create a plain JS API.  If you want to help make those things happen, that would be wonderful!

I noticed recently that someone has an atom editor setup for Clojure with several additions, including one for formatting Clojure/Clojurescript.  It would be great to have an easy way to integrate zprint with the atom editor in a similar way. 

I don't know what your use case is (and I would be interested in knowing), but whatever you want to do with zprint, if you want to help whack it into shape to be easier to use from JS, I'm all for it!

What specifically needs to happen to get an NPM package?  Likewise, (and more challenging, I suspect), what needs to happen to get zprint to have a more JS friendly API?

I'm more than happy to make the additions to the API as required, and run some process whenever I release a new version if you are willing to work on making the NPM build process happen and work with me on the JS API design that you think make sense.

**Comment by martinklepsch** [comment]
@roman01la I didn't start working on this. Mostly because I'm not very familiar with JS/NPM either and there are [other things](https://github.com/martinklepsch/clj-docs) I'm busy with right now.

@kkinnear the use case I mostly care about is editor integration. Modern editors like Atom and VS Code are comically bad at indenting Clojure code let alone format it. zprint could be useful to provide formatting, but as far as I understand it is not aimed at indentation. For this [shaunlebron/parlinter](https://github.com/shaunlebron/parlinter) might be more appropriate. Parlinter also is available as an NPM package so if anyone wants to just improve indentation for those editors it might be a better starting point. 

That said I do believe a formatting tool adds something useful to plain indentation and I'd love to see it becoming available as well.

**Comment by kkinnear** [comment]
No, you are quite correct, zprint isn't focused on just indentation.  It (generally) does the whole job, moving code and data from line to line as necessary to make it all more understandable.  I have added recent code to allow preserving newlines in vectors, to make hand-formatted hiccup and rum data come out looking better, but in general zprint ignores newlines.

**Comment by kkinnear** [comment]
To both of you, for what it is worth, I recently enhanced the `zprint-file-str` API to accept an options map.  This function takes a string (which is typically the contents of an entire file), and runs zprint on the entire string (file), including processing `;!zprint` directives.  This function already works in Clojurescript today, and I also fixed a bug where the `;!zprint` directives were leaking into the current options set (and now they don't).  This is the function that is used by `zprint-file` (and ultimately by `lein-zprint`), and does the heavy lifting for formatting multiple forms in a "file", or a string representing a file.  This is likely a key gateway function to any editor integration with zprint.

**Comment by martinklepsch** [comment]
That sounds promising @kkinnear — I guess a way forward could be 

- compiling zprint into a single JS file (:simple), checking if it all works
- setting `^:export` on `zprint-file-str` and compiling everything with :advanced
- check if everything is still working
- then it should be relatively easy to package that into a CLI script I guess

This "plan" does not address the NPM/JS API but could end up providing a minimal tool with fast startup that should be usable from editor extensions.

**Comment by kkinnear** [comment]
That's very helpful.  I've routinely test zprint with lumo, planck, and figwheel.  I don't _use_ it with any of those, however.  You've given me a good start.  My next step would be to try to find some other project that is doing what you suggest with a different code base, so I don't have to figure out how to do all of this from scratch.  My figwheel testing setup is probably close, but does both more and less than you are suggesting.  Is there some other project that (either of) you know about that does some (or all) of what you are suggesting for a different code base, that I could use as a guide?

**Comment by martinklepsch** [comment]
Unfortunately I’m not aware of such project. Maybe Roman has some further advice. I think shadow-cljs provides some particularly user friendly tools for cljs compilation these days so consider using it if you’re not too familiar with Figwheel or Boot.

**Comment by kkinnear** [comment]
Thanks, I'll start looking there.  @roman01la, if you have any ideas of similar projects or places to look for inspiration, I'd love to hear about them.  Or if you want to try your hand at this yourself, that would be fine too.

**Comment by roman01la** [comment]
I think I’ll take a look into all of these. Expect a PR later next week :)

**Comment by roman01la** [comment]
Here's some results:
- Advanced compiled build works as expected with `^:export` in Node.js
- Printing time for [cljs.core](https://github.com/clojure/clojurescript/blob/master/src/main/cljs/cljs/core.cljs) (320 KB, 11428 LOC) is ~22s for cold start and ~16s on repeated calls.

While I'm sure timing is not going to be that bad for common projects, it's interesting that Prettier (JS formatter) prints 500 KB minified JS file in ~500ms, [link](https://github.com/prettier/prettier/issues/443#issuecomment-276879173).

Perhaps we could improve this? I'll try to run the test through profiler and see where most of the time goes.

Since cold start seems to be 30% slower we could make use of a daemon process, [example](https://github.com/mantoni/eslint_d.js)

**Comment by martinklepsch** [comment]
That was quick! :)

I’m wondering - could file size / JS parsing be the issue here? 20s seems a little too much for just that but it could be a significant factor I guess?

How big was the advanced file? 

Maybe open a PR just for reference.

**Comment by roman01la** [comment]
Compiled zprint is around 450 KB. The startup time seems to be fine, it runs fast on smaller files. I think I just need to profile it and then we’ll see where to go next.

**Comment by thheller** [comment]
Hello everyone. Just found the discussion about this on Twitter and thought I can show off how `shadow-cljs` can help with this.

The repo is available here: https://github.com/thheller/zprint-npm

I already published it as a package to easily show how it would work. I published under the name `not-the-official-zprint` to mistakenly reserve any package name you may want to use.

Assuming you have the latest `npm` installed (with `npx`) you can just simply run:
```
cat some/file.clj | npx not-the-official-zprint
```

Otherwise you can install it globally and then run the command independenty

```
npm install -g not-the-official-zprint
cat some/file.clj | not-the-official-zprint
```

Compiling the entire thing was done with `shadow-cljs` running `shadow-cljs release script`.

Formatting the most minimal thing is very fast.

```
$echo "(foo)" | time node package/main.js
(foo)

        0.16 real         0.14 user         0.02 sys
```

**Comment by roman01la** [comment]
@thheller Unfortunately it doesn't make any difference, reformat time is still ~20s for 11kloc file

**Comment by martinklepsch** [comment]
@thheller I was considering pinging you, thanks for weighing in 👍

**Comment by martinklepsch** [comment]
@roman01la Could you post some more stats for different file sizes, i.e. 100loc, 500loc, 1kloc, 2kloc.
While it would definitely be nice to be fast with a 11kloc file these seem more typical. I hope this gives some idea whether it might already be "good enough" for some use cases.

**Comment by roman01la** [comment]
@martinklepsch You are right, will do. I'm just using it as a baseline.
@thheller Is there any potential perf improvements when building with shadow-clj?

**Comment by thheller** [comment]
`shadow-cljs` just helps with compiling/bundling, can't do anything to improve perf. Thats all left to the code and will be about the same regardless of which tool you use.

@roman01la didn't see the `zprint-clj` code you wrote until after I published mine. Looks more complete than mine although you committed blasphemy by writing parts of it in JS. :wink:

**Comment by roman01la** [comment]
@martinklepsch Some results for diff file size:

| LOC        | ~Time (s) |
| -------- |-------------| 
| 100        | 0.3              |
| 700        | 1.3   |
| 1500     | 3.2   |
| 11000   |  20 |

**Comment by martinklepsch** [comment]
I guess if we can get 1500loc down to `0.3` that could already be usable for most things. Oh and is this cold start or already running process?

**Comment by roman01la** [comment]
This was cold start results. Updated with repeated calls timings.

**Comment by roman01la** [comment]
@kkinnear Seems like [zprin.zprint/fzprint-hang-remaining](https://github.com/kkinnear/zprint/blob/master/src/zprint/zprint.cljc#L1622) takes the most time during reformat. Any clues what exactly can be slow there? From what I see in flame graph, it looks like there's no one slow function, but the algorithm itself.

**Comment by roman01la** [comment]
From the [line here](https://github.com/kkinnear/zprint/blob/master/src/zprint/zprint.cljc#L1649) it seems like `:hang` mode slows formatting a bit. Here's updated results with `:hang` mode disabled for all elements.

| LOC        | ~Time (s) |
| -------- |-------------| 
| 100        | 0.07              |
| 700        | 0.45  |
| 1500     | 1.45   |
| 11000   | 8.2 |

2x speedup!

**Comment by kkinnear** [comment]
Wow, you all have been busy!  

Regarding performance, zprint tries hard to do hang when it can -- where "when it can" is a complex function, including "not making it longer", and a variety of quite heuristic measures designed to replicate what a human might do to make a function look good.  Where good is "understandable".  To do this, it tries at each level to do a hang and a flow, and compares them.  Clearly, for deeply nested functions, this becomes exponentially costly.  With the JVM it will use multiple threads to help keep the cost down (though it didn't help that much, as I recall). 

I haven't had any opportunity to dig into the example you are using, but my experience is that some functions are extremely costly.  A classic in Clojure (and possibly Clojurescript, I haven't tried Clojurescript core sources) is the destructure function.  Takes >10s to czprint-fn it.  It isn't the length, it is the depth and how that depth interacts with the width (right margin) that causes the ballooning time growth.  

In the event that it is one or two functions causing the problem, one answer is to just turn off zprint for those functions using `;!zprint {:format :skip}`.  

Without hang support at all, zprint is going to be pretty uninteresting.  I have explored some ideas where zprint would reduce its efforts for the "best" formatting as it gets deeper into code.  It already does some of that when formatting deeply into maps, but not (as I recall) for code.

I want to look at the code that you are using as a test to see if there are any obviously difficult functions, and see what the JVM zprint does with it out of curiosity.

To answer your question -- `fzprint-hang-remaining` is pretty much what gets called after the beginning of a form is formatted.  If it is `:arg1` or `:binding` or whatever, after the beginning of the function (form) is formatted, `fzprint-hang-remaining` is called to do the rest of the function.  Which typically has, of course, other functions in it.

**Comment by roman01la** [comment]
Thanks! This is very helpful info.

> I want to look at the code that you are using as a test

I'm running it on [cljs.core](https://github.com/clojure/clojurescript/blob/master/src/main/cljs/cljs/core.cljs), this is where I'm getting ~20s result

**Comment by kkinnear** [comment]
I get about 10.6s doing that file with the zprint-filter (running on the JVM).  Probably about 1s of that is startup.

In looking at core.cljs itself (which I haven't done before, as I usually have been looking at Clojure core sources), I see a **lot** more protocols than I'm used to seeing.  Which I knew about, intellectually, but had no idea of the extent of their use.  I don't expect that is the source of the slowness, but it is a difference.

I didn't spend a lot of time looking at the file, but I did notice BitmapIndexedNode as a possible issue.  When I added a `;!zprint {:format :skip}` before that `deftype`, I saw about a .5s speedup.  Which isn't much, really.  I was hoping for one or two really awful functions.

The largest source in zprint is 3342 lines, and it takes about 6 seconds for the zprint-filter or lein-zprint to format it.  Since I do that only rarely, it doesn't bother me.  I didn't even know it took that long until I measured it just now.  Individual functions format, typically, in just over 1s (most of which is startup for the JVM), so I don't really notice that.  In part because I know that most of that is JVM startup, and that there isn't much more I can do about that (having already gotten it down to just under 1s).

It sounds like you have a useful ability to profile JS activity.  If you see anything that needs work, let me know.  Or if you can share how you do your JS profiling or point to someone else who has described it, I could use it to see what I can do to speed things up in Clojurescript.  I've spent a fair bit of time doing performance optimization, and at least on the JVM I think I've gotten most of the low hanging fruit and a good bit of the next level as well.  But there is always more to do in performance optimization -- and it only takes one good idea to make a huge difference.

**Comment by kkinnear** [comment]
Having had a chance to look at what you all did, I'm even more impressed!

Here are some comparison timings, averaged over three runs each (all in seconds):

Tool | core.cljs (11428 loc) | core.cljs {:list {:hang? false}} | core.cljs no hang at all | zprint.cljc (3342 loc)
---- | --------- | ---------------------------- | ----------------- | --------------------
zprint-filter-0.4.6 (JVM) | 10.4 | 9.4 | 8.4 | 6.35
lumo zprint filter | 43.7 | | | 19.46
zprint-clj (npm) | 18.9 | | | 5.6

Presumably the compiler difference is what drives the lumo numbers down.

I can't quite tell if you've got this integrated enough into atom (or whatever) to be able to tell if it is fast enough to be useful or not.  I'm gradually (slowly) coming up to speed on node but as yet I don't have any tools to profile things there, to see if there are things I could do to make it better.  I suppose the question for you all is: is it useful as it stands, or does it **have** to get faster to be useable?  If so, how **much** faster?

Thanks for all of your efforts!

**Comment by kkinnear** [comment]
Here is an updated table of timings.  I didn't realize that zprint-clj defaulted to no hang at all.  I've also tried `not-the-official-zprint`, and the results were pretty much the same as zprint-clj (with `--hang` specified).  Now, `lumo` doesn't look like such an outlier.  Still slower, but not many times slower.

Tool | core.cljs (11428 loc) | core.cljs {:list {:hang? false}} | core.cljs no hang at all | zprint.cljc (3342 loc) | fzprint-two-up (280 loc)
---- | --------- | ---------------------------- | ----------------- | -------------------- | ------------------------
zprint-filter-0.4.6 (JVM) | 10.4 | 9.4 | 8.4 | 6.35 | 1.9
lumo zprint filter | 43.7 | | | 19.46 | 2.6
zprint-clj (npm) | 37.4 | | 18.9 | 16.9 | 1.6

This matches my previous benchmarking much more closely.  Node starts up much faster then the JVM, making it faster on small things, but the JVM wins when things take a long time.  That said, I think that most usage is for small things.  The new right hand column is a rather large 280 loc function.  Most functions aren't this large, so that startup time dominates almost everything.

**Comment by roman01la** [comment]
Interesting... I agree that most thing you'd want to format in the editor are small files, unless you are working with Clojure or ClojureScript core for example.

I've recorded a short screencast explaining how to profile zprint-clj in Chrome DevTools, in case if you want to take a look and find out how to improve performance https://www.youtube.com/watch?v=c5wDQNMk0BY
Chekout zprint-clj repo, start local server from the root of the repo and follow the screencast

**Comment by kkinnear** [comment]
I'm about to release 0.4.7, in which I worked hard to try to get the performance better.  I did a lot of profiling, in both cljs and clj, and rewrote one internal algorithm iteratively, for a small performance gain.  I managed to reduce the time spent on very deeply nested and large functions, but that isn't a usual situation, and didn't help the core.cljs time much.  I tried several algorithmic changes, recognizing that the fastest function call is the one you don't make, but I didn't find anything that made a significant difference that didn't ruin the formatting too much. I did find a bug where I wasn't running multithreaded in lein-zprint, for a 20% improvement there, but that's not anything that helps out the node performance.  I spent all of the time I have had to spend in the last month on performance, and so I'm going to move on now (though there are always interesting things to try).
